src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostDocumentCompletionEndpoint.cs

```diff
@@ -0,0 +1,297 @@
+        return [];
+    }
+
+    protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(RoslynCompletionParams request)
+        => request.TextDocument is null ? null : RoslynLspExtensions.ToRazorTextDocumentIdentifier(request.TextDocument);
+
+    protected override Task<VSInternalCompletionList?> HandleRequestAsync(RoslynCompletionParams request, RazorCohostRequestContext context, CancellationToken cancellationToken)
+        => HandleRequestAsync(request, context.TextDocument.AssumeNotNull(), cancellationToken);
+
+    private async Task<VSInternalCompletionList?> HandleRequestAsync(RoslynCompletionParams request, TextDocument razorDocument, CancellationToken cancellationToken)
+    {
+        if (request.Context is null || ToVsLSP<VSInternalCompletionContext>(request.Context) is not VSInternalCompletionContext completionContext)
+        {
+            _logger.LogError("Completion request context is null");
+            return null;
+        }
+
+        // Return immediately if this is auto-shown completion but auto-shown completion is disallowed in settings
+        var clientSettings =  _clientSettingsManager.GetClientSettings();
+        var autoShownCompletion = completionContext.TriggerKind != CompletionTriggerKind.Invoked;
+        if (autoShownCompletion && !clientSettings.ClientCompletionSettings.AutoShowCompletion)
+        {
+            return null;
+        }
+
+        _logger.LogDebug($"Invoking completion for {razorDocument.FilePath}");
+
+        if (await _remoteServiceInvoker.TryInvokeAsync<IRemoteCompletionService, CompletionPositionInfo?>(
+                razorDocument.Project.Solution,
+                (service, solutionInfo, cancellationToken)
+                    => service.GetPositionInfoAsync(
+                            solutionInfo,
+                            razorDocument.Id,
+                            completionContext,
+                            ToVsLSP<Position>(request.Position).AssumeNotNull(),
+                            cancellationToken),
+                cancellationToken).ConfigureAwait(false) is not { } completionPositionInfo)
+        {
+            // If we can't figure out position info for request position we can't return completions
+            return null;
+        }
+
+        var documentPositionInfo = completionPositionInfo.DocumentPositionInfo;
+        if (documentPositionInfo.LanguageKind != RazorLanguageKind.Razor)
+        {
+            completionContext = DelegatedCompletionHelper.RewriteContext(completionContext, documentPositionInfo.LanguageKind);
+        }
+
+        // First of all, see if we in HTML and get HTML completions before calling OOP to get Razor completions.
+        // Razor completion provider needs a set of existing HTML item labels.
+
+        VSInternalCompletionList? htmlCompletionList = null;
+        var razorCompletionOptions = new RazorCompletionOptions(
+            SnippetsSupported: true, // always true in non-legacy Razor
```

Comment:
This comment would seem to back up the idea that `SnippetsSupported` can simply be removed entirely

Reply:
Not quite, always true in **non-legacy Razor**, but also used in **legacy Razor** where it's not true. :)

Reply:
Yeah, I was thinking "legacy" in this case meant the legacy completion stuff you got rid of a while ago. I forgot about the legacy editor stuff.